### PR TITLE
Remove copies for LayeredImage creation

### PR DESCRIPTION
**Pull request by jeremykubica**
When given a set of `RawImage`s the `LayeredImage` constructor does a copy. This PR adds a constructor that allows `LayeredImage` to take in the underlying Eigen data array (or numpy array from Python), claim ownership of the memory, and avoid the copy.

We also use `ImageStack`'s new `append_image()` function to add images to the stack one on one instead of allocating a full list and copying them in. This slightly reduces the peak memory needed. We include a `force_move` option that allows the `ImageStack` to take ownership of the image and avoid doing a copy when it inserts it into the vector.

### CODE REVIEW

File: src/kbmod/reprojection.py [hunk 1]
```diff
@@ -287,29 +281,26 @@ def _reproject_work_unit_in_parallel(
     # when all the multiprocessing has finished, convert the returned numpy arrays to RawImages.
     concurrent.futures.wait(future_reprojections, return_when=concurrent.futures.ALL_COMPLETED)
     time_stamps = []
```
Comment (drewoldag) [inline src/kbmod/reprojection.py:283]: I think that we can remove this variable entirely, if we're able to use the `ImageStack.sort_by_time()` method.

File: src/kbmod/reprojection.py [hunk 2]
```diff
@@ -287,29 +281,26 @@ def _reproject_work_unit_in_parallel(
...
-            mask_raw_image,
+            science_add,
+            variance_add,
+            mask_add,
             psf,
+            time,
         )
 
         # append timestamps and layeredImages to lists
         time_stamps.append(time)
```
Comment (drewoldag) [inline src/kbmod/reprojection.py:299]: And can remove this line for the same reason as above.

File: src/kbmod/search/layered_image.cpp [hunk 3]
```diff
@@ -306,6 +316,9 @@ static void layered_image_bindings(py::module& m) {
 
     py::class_<li>(m, "LayeredImage", pydocs::DOC_LayeredImage)
             .def(py::init<const ri&, const ri&, const ri&, pf&>())
+            .def(py::init<search::Image&, search::Image&, search::Image&, pf&, double>(),
+                 py::arg("sci").noconvert(true), py::arg("var").noconvert(true),
+                 py::arg("msk").noconvert(true), py::arg("psf"), py::arg("obs_time"))
```
Comment (drewoldag) [inline src/kbmod/search/layered_image.cpp:321]: What do these `noconvert`s do? Just curious, it's unfamiliar.

**Review by drewoldag — APPROVED** [review]
This looks great, thanks! No major concerns, mostly just happy to see things getting cleaned up.

File: src/kbmod/reprojection.py (hunk 1 above)
Comment (jeremykubica) [reply src/kbmod/reprojection.py]: Cleaned it up

File: src/kbmod/reprojection.py (hunk 2 above)
Comment (jeremykubica) [reply src/kbmod/reprojection.py]: cleaned this up too.

File: src/kbmod/search/layered_image.cpp (hunk 3 above)
Comment (jeremykubica) [reply src/kbmod/search/layered_image.cpp:321]: I don't know all the details. But they are needed for passing numpy arrays of float32s into Eigen matrices in C++. I'm not sure if it prevents it from converting to a float64 or into a list of lists or something else.

**Pull request merged**